### ISSUE TITLE
V2 Matchers don't have ability to match

### ISSUE THREAD

**Issue by Zertsov** [prometheus/alertmanager#2766]
This may be completely on me, but I cannot find anywhere where v2 Matchers have the ability to actually _match_ anything, as compared to the old Prometheus labels matchers that were used in the v1 client.

**What did you do?**
Attempted to migrate v1 Golang client application to v2 and use Matchers for Silence responses.

**What did you expect to see?**
Matchers field within a Silence to have the ability to match regex.

**What did you see instead? Under which circumstances?**
V2 Matchers within Silence have no ability to match.

**Environment**

* System information:

	Darwin 20.6.0 x86_64

* Alertmanager version:

	insert output of `alertmanager --version` here (repeat for each alertmanager
	version in your cluster, if relevant to the issue)

* Prometheus version:

	insert output of `prometheus --version` here (repeat for each prometheus
	version in your cluster, if relevant to the issue)

* Alertmanager configuration file:
```
insert configuration here
```

* Prometheus configuration file:
```
insert configuration here (if relevant to the issue)
```

* Logs:
```
insert Prometheus and Alertmanager logs relevant to the issue here
```

**Comment by beorn7** [comment]
Could you provide more detailed information so that we can easily produce the bug?

Like concrete request and response, old and new, and what precisely you were expecting to match.

**Comment by sinkingpoint** [comment]
What I think this is referring to is that the v1 client (formerly api/client.go) returned `label.Matcher`s which have a `Matches` method https://github.com/prometheus/alertmanager/blob/fd0929ba9fc58737a9c91f24771862692fa72d17/pkg/labels/matcher.go#L81

Conversely, the v2 client returns the auto-generated `models.Matcher` which lacks the same methods for actually using those to match strings. You can certainly manually transform one in to the other, but doing so is clunky at best

**Comment by Zertsov** [comment]
sorry for the late comment, but as @sinkingpoint said, the matchers used in the V1 client have the type `labels.Matcher` which have different properties and methods that allow users to do matching with things like regex, while the generated matchers in the V2 client are missing properties and methods that allow for actually checking if something matches or not.

https://github.com/prometheus/alertmanager/blob/main/api/v2/models/matcher.go#L32

vs

https://github.com/prometheus/alertmanager/blob/fd0929ba9fc58737a9c91f24771862692fa72d17/pkg/labels/matcher.go#L81

**Comment by beorn7** [comment]
I might not be the real expert for this area, but I would say the generated matchers for the v2 API are pure DTOs, just so that you can construct the payload for an API call with them. They are not supposed to have any methods actually acting on the contained data.

If you want the actual matcher functionality, you have to use the "real" matchers. (But why would you? Alertmanager is supposed to do the matching. A client just sends the matchers to Alertmanager via the API.)

**Comment by sinkingpoint** [comment]
You are correct that they are DTOs, but there's plenty of usecases for wanting to use Matchers themselves (one of mine in particular is performing adhoc analysis on silences before they're applied - working out how many alerts a silence would silence etc). I'm in favor of leaving the swagger generated stuff alone, but would you tick a PR that added some sort of convenience method to convert between the types?

Something like `labels.NewMatcherFromDTO(models.Matcher) labels.Matcher`?

**Comment by beorn7** [comment]
I thought something like this must already exist in the codebase. I'm not an expert in the Swagger stuff, but since Alertmanager can tell you about matchers in its API and it accepts matchers via its API, the conversion must be happening somewhere, doesn't it?

In any case, I would leave it to a real expert of this code to decide what can be done here.

**Comment by Zertsov** [comment]
Ended up making this work by grabbing the payload out of `GetSilenceOK` and then marshalling the `GettableSilences` and then unmarshalling those into `types.Silence`. Allowed me to do the analysis of the silences that I was doing beforehand.

**Comment by TheMeier** [comment]
@Zertsov is this issue still relevant?
